File: src/main/python/solution_day2.py
import os
import file_reader
from enum import Enum


class Strategy(Enum):
    Rock = 1
    Paper = 2
    Scissor = 3



def main():
    raw_data = file_reader.read_file_and_split("../source/day2/rock_paper_scissor", os.linesep)
    strategies = transform_to_arr_of_strategy(raw_data)
    result = sum([calculate_result(match) for match in strategies])
    print("Result of following strategies:", result)
    # cleaned_transformed_data = transform_to_arr(raw_data)
    # print_result(cleaned_transformed_data)


def transform_to_arr_of_strategy(data):
    return [(string_to_strategy(line[0]), string_to_strategy(line[2])) for line in data]


def string_to_strategy(strategy_string):
    match strategy_string:
        case "A" | "X":
            return Strategy.Rock
        case "B" | "Y":
            return Strategy.Paper
        case "C" | "Z":
            return Strategy.Scissor


def calculate_result(matchup):
    return match_to_result(matchup) + matchup[1].value


def match_to_result(matchup):
    win = 6
    draw = 3
    loss = 0

    match matchup:
        case (Strategy.Rock, Strategy.Paper):
            return win
        case (Strategy.Rock, Strategy.Scissor):
            return loss
        case (Strategy.Paper, Strategy.Scissor):
            return win
        case (Strategy.Paper, Strategy.Rock):
            return loss
        case (Strategy.Scissor, Strategy.Rock):
            return win
        case (Strategy.Scissor, Strategy.Paper):
            return loss
        case _:
            return draw


main()

File: src/main/scala/adventofcode/day2/SolutionDay2.scala
package adventofcode.day2

import adventofcode.day2.SolutionDay2Data.{DRAW, LOSS, Paper, Rock, Scissor, Strat, WIN}
import util.FileUtil.readFileAndSplitBy

object SolutionDay2 extends App {

  val fileInputDelimited = readFileAndSplitBy("src/main/source/day2/rock_paper_scissor", "\n")

  val transformed: List[(Strat, Strat)] = fileInputDelimited.toList.map(line => toStrategyTuple(line.split(" ")))
  val transformedV2: List[(Strat, Strat)] = fileInputDelimited.toList.map(line => toStrategyTupleV2(line.split(" ")))

  val result = transformed.map(determinePoints)
  val resultV2 = transformedV2.map(determinePoints)

  println("If all goes well: " + result.sum)
  println("If all goes well V2: " + resultV2.sum)

  def determinePoints(matchUp: (Strat, Strat)): Int = {
    val matchResult = matchUp match {
      case (Rock, Paper) => WIN
      case (Rock, Scissor) => LOSS
      case (Paper, Scissor) => WIN
      case (Paper, Rock) => LOSS
      case (Scissor, Rock) => WIN
      case (Scissor, Paper) => LOSS

      case (_, _) => DRAW // other scenarios
    }
    matchResult + matchUp._2.value
  }

  def toStrategyTuple(matchup: Array[String]): (Strat, Strat) = {
    matchup match {
      case Array(x, y) => (toStrategy(x), toStrategy(y))
    }
  }

  def toStrategy(strategy: String): Strat = {
    strategy match {
      case "A" | "X" => Rock
      case "B" | "Y" => Paper
      case "C" | "Z" => Scissor
    }
  }

  def toStrategyTupleV2(matchup: Array[String]): (Strat, Strat) = {
    val opponent: Strat = toStrategy(matchup.head)
    val strategy: Strat = toStrategyV2(opponent, matchup.tail.head)

    (opponent, strategy)
  }

  def toStrategyV2(opponent: Strat, strategy: String): Strat = {
    (opponent, strategy) match {
      case (Rock, "X") => Scissor
      case (Rock, "Y") => Rock
      case (Rock, "Z") => Paper

      case (Paper, "X") => Rock
      case (Paper, "Y") => Paper
      case (Paper, "Z") => Scissor

      case (Scissor, "X") => Paper
      case (Scissor, "Y") => Scissor
      case (Scissor, "Z") => Rock
    }
  }
}

object SolutionDay2Data {
  val WIN: Int = 6
  val DRAW: Int = 3
  val LOSS: Int = 0

  trait Strat {
    def value: Int
  }

  case object Rock extends Strat {
    override def value: Int = 1
  }

  case object Paper extends Strat {
    override def value: Int = 2
  }

  case object Scissor extends Strat {
    override def value: Int = 3
  }
}
